File: iosApp/iosApp/Pages/StopDetails/StopDetailsRouteView.swift
//
//  StopDetailsRouteView.swift
//  iosApp
//
//  Created by Horn, Melody on 2024-04-03.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Foundation
import Shared
import SwiftUI

struct StopDetailsRouteView: View {
    var analytics: Analytics = AnalyticsProvider.shared
    let patternsByStop: PatternsByStop
    let now: Instant
    let pushNavEntry: (SheetNavigationStackEntry) -> Void
    let pinned: Bool
    let onPin: (String) -> Void

    var body: some View {
        if let line = patternsByStop.line {
            LineCard(line: line, routes: patternsByStop.routes, pinned: pinned, onPin: onPin) {
                StopDeparturesSummaryList(
                    patternsByStop: patternsByStop,
                    condenseHeadsignPredictions: patternsByStop.routes.count > 1,
                    now: now,
                    context: .stopDetailsUnfiltered,
                    pushNavEntry: pushNavEntry,
                    analytics: analytics,
                    pinned: pinned
                )
            }
        } else if let route = patternsByStop.routes.first {
            RouteCard(route: route, pinned: pinned, onPin: onPin) {
                StopDeparturesSummaryList(
                    patternsByStop: patternsByStop,
                    condenseHeadsignPredictions: false,
                    now: now,
                    context: .stopDetailsUnfiltered,
                    pushNavEntry: pushNavEntry,
                    analytics: analytics,
                    pinned: pinned
                )
            }
        } else {
            EmptyView()
        }
    }
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/stopDetails/StopDetailsRouteView.kt
package com.mbta.tid.mbta_app.android.stopDetails

import androidx.compose.runtime.Composable
import com.mbta.tid.mbta_app.android.component.legacyRouteCard.LegacyRouteCard
import com.mbta.tid.mbta_app.android.component.legacyRouteCard.LineCard
import com.mbta.tid.mbta_app.android.component.legacyRouteCard.StopDeparturesSummaryList
import com.mbta.tid.mbta_app.model.PatternsByStop
import com.mbta.tid.mbta_app.model.RealtimePatterns
import com.mbta.tid.mbta_app.model.StopDetailsFilter
import com.mbta.tid.mbta_app.model.TripInstantDisplay
import kotlinx.datetime.Instant

@Composable
fun StopDetailsRouteView(
    patternsByStop: PatternsByStop,
    now: Instant,
    pinned: Boolean,
    onPin: (String) -> Unit,
    updateStopFilter: (StopDetailsFilter?) -> Unit
) {
    fun onTappedPatterns(patterns: RealtimePatterns) {
        updateStopFilter(StopDetailsFilter(patternsByStop.routeIdentifier, patterns.directionId()))
    }
    if (patternsByStop.line != null) {
        LineCard(patternsByStop.line!!, patternsByStop.routes, pinned, onPin) {
            StopDeparturesSummaryList(
                patternsByStop,
                condenseHeadsignPredictions = patternsByStop.routes.size > 1,
                now,
                TripInstantDisplay.Context.StopDetailsUnfiltered,
                pinned,
                onClick = ::onTappedPatterns
            )
        }
    } else {
        val route = patternsByStop.routes.firstOrNull() ?: return
        LegacyRouteCard(route, pinned, onPin) {
            StopDeparturesSummaryList(
                patternsByStop,
                false,
                now,
                TripInstantDisplay.Context.StopDetailsUnfiltered,
                pinned,
                onClick = ::onTappedPatterns
            )
        }
    }
}
